### DISCUSSION TITLE
[@types/antlr4] Import/Export paths changed in antlr4 4.8.1

### DISCUSSION THREAD

**Discussion by mint-thompson** [DefinitelyTyped/DefinitelyTyped#56220] [Issues with a @types package] [answered]
Creating this discussion to continue from #49798. Some of the exports from the antlr4 package have been restructured as of the 4.8.1 release. For example, `error` is now available on the main import, instead of requiring an import of `antlr4/error`, and many classes such as `FileStream` and `LL1Analyzer` are newly available as exports. As a result, the 4.7.2 release of @types/antlr4 does not work well with antlr4 versions 4.8.1 and up. Is anyone currently working on updated types for antlr4? If not, I may be able to find time to work on it, but I can't guarantee that.

EDIT: Although these types should certainly still be updated, I was able to make a workaround that allowed me to keep using the current type definitions. See this discussion comment: https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/56220#discussioncomment-1495658

EDIT 2022/03/29: The workaround is not quite as convenient as I initially thought, but may still be useful if you don't mind adding in another library for module resolution at runtime.

[a comment by an automated account is not shown]

**Comment by tarilabs** [comment, marked as answer]
>  Is anyone currently working on updated types for antlr4?

Not me, I believe the bot involved me as one of the past committers here, as I fixed something in the past; like in other circumstances, I'm happy to support/test in my capacity, but I have no bandwidth personally to work on this.

@mint-thompson I would encourage you to give a shot at it since you seem to have already identified a lot of potential fixes and improvements; don't hesitate to ping around, the DefinitelyTyped community is very supportive I found in my (limited) experience 🚀 

FYI you might also want to check out some previous conversations here https://github.com/antlr/antlr4/issues/2968 for headsup.

**Reply by mint-thompson** [reply]
Okay, I'm going to give it a try. Thank you for the link to the issue on the antlr4 repo.

**Comment by mint-thompson** [comment]
I wanted to add some information to this thread that may be useful to folks who are also trying to use newer versions of antlr4 in their TypeScript projects. I was able to get my project working with antlr4 version 4.9.2 and the existing type definitions by making these changes:
- in `tsconfig.json`, I added two mappings to `compilerOptions.paths`:
```
"antlr4": ["node_modules/antlr4/src/antlr4"],
"antlr4/*": ["node_modules/antlr4/src/antlr4/*"]
```
- in places where I imported generated classes, I changed destructured imports to direct imports, since the new generated classes are exported using `export default`. For example, for the generated parser class, `import { MyParser } from './generated/MyParser';` is changed to `import MyParser from './generated/MyParser';`.
- since the project uses jest for testing, I also added this to `jest.config.js`:
```
moduleNameMapper: {
  '^antlr4(.*)$': '<rootDir>/node_modules/antlr4/src/antlr4$1'
}
```
- I also modified the `transform` option in `jest.config.js` to have `ts-jest` operate on `js` files. If you're using a different transformer, you may need to do something different.

With these changes, I didn't have to modify any of the imports of the antlr4 runtime.

**Reply by mint-thompson** [reply]
Adding some further information for anyone who tried adding to paths: this, unfortunately, doesn't actually work all that well unless you also use some other library for resolving imports (such as tsconfig-paths), since the module resolution paths that are given in `tsconfig.json` aren't known to node at runtime. So, the situation has returned to a case where the types are in definite need of an update.
